compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Repl.hs

```diff
@@ -154,12 +154,31 @@ topologicalSort lfPkgs = map toPkg $ topSort $ transposeG graph
     toPkg = (\(pkg, _, _) -> pkg) . fromVertex
 
 data ReplState = ReplState
-  { moduleNames :: [LF.ModuleName]
+  { imports :: [ImportDecl GhcPs]
   , bindings :: [(LPat GhcPs, Type)]
   , lineNumber :: Int
   }
 
-type ReplM a = Repl.HaskelineT (State.StateT ReplState IO) a
+type ReplM = Repl.HaskelineT (State.StateT ReplState IO)
+
+data ReplInput
+  = ReplStatement (Stmt GhcPs (LHsExpr GhcPs))
+  | ReplImport (ImportDecl GhcPs)
+
+parseReplInput :: String -> DynFlags -> Either Error ReplInput
+parseReplInput input dflags = swapEither $ do
+    -- We use @Either ReplInput Error@ to short-circuit on success.
+    -- The most common input will be statements. So, we try parsing statements
+    -- first and emit statement parse errors on failure.
+    e <- tryParse (parseStatement input dflags) (ReplStatement . unLoc)
+    _ <- tryParse (parseImport input dflags) (ReplImport . unLoc)
+    pure e
+  where
+    swapEither = either Right Left
+    tryParse :: ParseResult a -> (a -> ReplInput) -> Either ReplInput Error
```

Comment:
Slightly disappointed that neither `ParseResult` nor `P` have alternative instances or anything close to it.

Reply:
Yeah, my thoughts exactly :disappointed: